### PR TITLE
fix: Handle selecting files and folders for patch options correctly

### PR DESCRIPTION
**Pull request by Francesco146**
closes https://github.com/ReVanced/revanced-manager/issues/1703

### CODE REVIEW

File: pubspec.yaml [hunk 1]
```diff
@@ -23,6 +23,7 @@ dependencies:
   dynamic_color: ^1.7.0
   dynamic_themes: ^1.1.0
   expandable: ^5.0.1
+  file_picker: ^8.0.5
```
Comment (Francesco146) [inline pubspec.yaml:26]: to use the latest version we need to ubump the minimum Android SDK version to 21

Comment (oSumAtrIX) [reply pubspec.yaml:26]: what is it right now/what android version does it correspond to

Comment (Francesco146) [reply pubspec.yaml:26]: not sure, but i tried bumping it to the latest but in the next version they dropped support and the build fails

https://pub.dev/packages/file_picker/changelog


I'll take a deeper look soon enough

**Comment by TheAabedKhan** [comment]
I am out of town right now. Wait for around 15 hours please.

File: pubspec.yaml (hunk 1 above)
Comment (Francesco146) [reply pubspec.yaml:26]: @oSumAtrIX appears to be Android SDK 19 (A4.4)

File: lib/ui/widgets/patchesSelectorView/patch_options_fields.dart [hunk 2]
```diff
@@ -524,29 +528,53 @@ class _TextFieldForPatchOptionState extends State<TextFieldForPatchOption> {
...
-                    case 'file':
-                      final String? result = await FlutterFileDialog.pickFile();
-                      if (result != null) {
-                        controller.text = result;
-                        widget.onChanged(controller.text);
+                  // manageExternalStorage permission is required for file/folder selection
+                  // otherwise, the app will not complain, but the patches will error out
+                  // the same way as if the user selected an empty file/folder.
+                  // Android 11 and above requires the manageExternalStorage permission
+                  final Map<String, dynamic> availableActions = {
```
Comment (TheAabedKhan) [inline lib/ui/widgets/patchesSelectorView/patch_options_fields.dart:535]: Reduce code duplication
```suggestion
                  if (selection == 'file' || selection == 'folder') {
                    if (_managerAPI.isScopedStorageAvailable) {
                      final permission =
                          await Permission.manageExternalStorage.request();
                      if (!permission.isGranted) {
                        return;
                      }
                    }
                  }
```

**Review by TheAabedKhan — CHANGES_REQUESTED** [review]
Use `switch-case` block instead of `Map` as it is  potentially faster when dealing with a small, fixed set of cases due to compile-time optimizations.

```dart
switch (selection) {
  case 'file':
    final FilePickerResult? result =
        await FilePicker.platform.pickFiles();
    if (result == null) {
      return;
    }
    controller.text = result.files.single.path!;
    widget.onChanged(controller.text);
    break;
  case 'folder':
    final String? result =
        await FilePicker.platform.getDirectoryPath();
    if (result == null) {
      return;
    }
    controller.text = result;
    widget.onChanged(controller.text);
    break;
  case 'remove':
    widget.removeValue!();
    break;
  case 'null':
    controller.text = '';
    widget.onChanged(null);
    break;
}
```

**Comment by Francesco146** [comment]
will do within 10h

**Comment by kitadai31** [comment]
When select a file with `FilePicker.platform.pickFiles();`, the selected file will be copied to the cache directory of the app.
So there is no need to request the manage all files permission if "Select file path" is pressed.

(Ideally, manager should get the file path directly without copying, though.)

**Comment by Francesco146** [comment]
@kitadai31 I'll remove the check for all permissions in pickFiles, but idk if the second approach is doable rn

**Review by TheAabedKhan — APPROVED** [review]

**Pull request merged**